server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4515,26 +4517,37 @@ public void syncSegmentsFromRemoteSegmentStore(boolean overrideLocal, boolean re
                     }
                     storeDirectory.copyFrom(remoteDirectory, file, file, IOContext.DEFAULT);
                     downloadedSegments.add(file);
-                    if (file.startsWith(SEGMENT_INFO_SNAPSHOT_FILENAME_PREFIX)) {
-                        assert segmentInfosSnapshotFilename == null : "There should be only one SegmentInfosSnapshot file";
-                        segmentInfosSnapshotFilename = file;
-                    }
                 } else {
                     skippedSegments.add(file);
                 }
             }
-            if (refreshLevelSegmentSync && segmentInfosSnapshotFilename != null) {
+
+            if (refreshLevelSegmentSync && remoteSegmentMetadata != null) {
                 try (
                     ChecksumIndexInput indexInput = new BufferedChecksumIndexInput(
-                        storeDirectory.openInput(segmentInfosSnapshotFilename, IOContext.DEFAULT)
-                    )
+                        new ByteArrayIndexInput("Snapshot of SegmentInfos", remoteSegmentMetadata.getSegmentInfosBytes())
+                    );
                 ) {
                     SegmentInfos infosSnapshot = SegmentInfos.readCommit(
                         store.directory(),
                         indexInput,
-                        Long.parseLong(segmentInfosSnapshotFilename.split("__")[1])
+                        remoteSegmentMetadata.getGeneration()
                     );
                     long processedLocalCheckpoint = Long.parseLong(infosSnapshot.getUserData().get(LOCAL_CHECKPOINT_KEY));
+                    // Following code block makes sure to commit SegmentInfosSnapshot with the highest generation.
+                    // If local filesystem already has segments_N+2 and infosSnapshot has generation N, after commit,
+                    // there would be 2 files that would be created segments_N+1 and segments_N+2. With the policy of preserving
+                    // only the latest commit, we will delete segments_N+1 which in fact is the part of the latest commit.
+                    Optional<String> localMaxSegmentInfos = localSegmentFiles.stream()
+                        .filter(file -> file.startsWith(IndexFileNames.SEGMENTS))
+                        .max(Comparator.comparingLong(SegmentInfos::generationFromSegmentsFileName));
+                    if (localMaxSegmentInfos.isPresent()
+                        && infosSnapshot.getGeneration() < SegmentInfos.generationFromSegmentsFileName(localMaxSegmentInfos.get()) - 1) {
+                        SegmentInfos localSegmentInfos = store.readLastCommittedSegmentsInfo();
+                        if (localSegmentInfos.files(false).equals(infosSnapshot.files(false))) {
+                            infosSnapshot = localSegmentInfos.clone();
+                        }
+                    }
```

Comment:
should we assert that `localSegmentInfos` corresponds to the files we download from remote ?

Reply:
We need to check on the `else` part here as well . Ideally the `else` should never be hit. So we can thrown an exception in that case.

Reply:
Also can we explore if we can delete the local `segments_N+2` and start fresh post the refresh ?

Reply:
Deleting worked. Thanks @gbbafna